stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+
+    [unzip xs] is equivalent to [(map fst xs, map snd xs)].
+
+    Querying either of the sequences returned by [unzip xs]
+    causes [xs] to be queried.
+    Therefore, querying both of them
+    causes [xs] to be queried twice.
+    Thus, [xs] must be persistent and cheap.
+    If that is not the case, use [unzip (memoize xs)].
+
+    @since 4.14 *)
+
+val partition_either : ('a, 'b) Either.t t -> 'a t * 'b t
+(** [partition_either] transforms a sequence of fruit (either apples or oranges)
+    into a pair of a sequence of apples and a sequence of oranges.
+
+    [partition_either xs] is equivalent to
+    [(filter_map Either.find_left xs, filter_map Either.find_right xs)].
+
+    Querying either of the sequences returned by [partition_either xs]
+    causes [xs] to be queried.
+    Therefore, querying both of them
+    causes [xs] to be queried twice.
+    Thus, [xs] must be persistent and cheap.
+    If that is not the case, use [partition_either (memoize xs)].
+
+    @since 4.14 *)
+
+val partition : ('a -> bool) -> 'a t -> 'a t * 'a t
+(** [partition p xs] returns a pair of the subsequence of the elements
+    of [xs] that satisfy [p] and the subsequence of the elements of
+    [xs] that do not satisfy [p].
+
+    [partition p xs] is equivalent to
+    [filter p xs, filter (fun x -> not (p x)) xs].
+
+    Consuming both of the sequences returned by [partition p xs] causes
+    [xs] to be consumed twice and causes the function [f] to be applied
+    twice to each element of the list.
+    Therefore, [f] should be pure and cheap.
+    Furthermore, [xs] should be persistent and cheap.
+    If that is not the case, use [partition p (memoize xs)].
+
+    @since 4.14 *)
+
+(** {1 Converting between sequences and iterators} *)
+
+(** An iterator is a representation of a sequence as a function of type
+    [unit -> 'a option]. Every time this function is invoked, it returns
+    the next element of the sequence. When there are no more elements,
+    it returns [None]. An iterator has mutable internal state, therefore
+    is ephemeral: the sequence that it represents can be consumed at most
+    once. *)
+
+val of_iterator: (unit -> 'a option) -> 'a t
+(** [of_iterator it] is the sequence of the elements produced by the iterator
+    [it]. It is an ephemeral sequence: it can be consumed at most once. If a
+    persistent sequence is needed, use [memoize (of_iterator it)].
+    @since 4.14 *)
+
+val to_iterator: 'a t -> (unit -> 'a option)
+(** [to_iterator xs] is a fresh iterator on the sequence [xs].
+    @since 4.14 *)
+
+(** {1 Sequences of integers} *)
+
+val ints: int -> int t
+(** [ints i] is the infinite sequence of the integers, beginning at [i] and
+    counting up.
+    @since 4.14 *)
+
+val up: int -> int -> int t
+(** [up i j] is the finite sequence of the integers comprised between
+    [i] included and [j] excluded, counting up. It is nonempty if and
+    only if [i < j] holds.
+    @since 4.14 *)
```

Comment:
There was an extensive, inconclusive discussion of a similar function in #10480.  Did you read it?

I find the name `Seq.range` more evocative than `Seq.up`.  Not sure `Seq.down` is much useful.

Also: why having these functions at type `int` but not at types `int32`, `int64`, `float`, ... ?

Reply:
I'll remove them and leave the decision for range to #10480 . No need to duplicate efforts.